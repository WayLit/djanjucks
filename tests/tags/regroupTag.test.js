import djanjucks from '../../src';

const CITIES = [
  { name: 'Mumbai', population: '19,000,000', country: 'India' },
  { name: 'Calcutta', population: '15,000,000', country: 'India' },
  { name: 'New York', population: '20,000,000', country: 'USA' },
  { name: 'Chicago', population: '7,000,000', country: 'USA' },
  { name: 'Tokyo', population: '33,000,000', country: 'Japan' }
];

const CITIES_WITH_OBJECT = [
  {
    name: 'Mumbai',
    population: '19,000,000',
    country: { name: 'India', color: 'red' }
  },
  {
    name: 'Calcutta',
    population: '15,000,000',
    country: { name: 'India', color: 'blue' }
  },
  {
    name: 'New York',
    population: '20,000,000',
    country: { name: 'USA', color: 'yellow' }
  },
  {
    name: 'Chicago',
    population: '7,000,000',
    country: { name: 'USA', color: 'red' }
  },
  {
    name: 'Tokyo',
    population: '33,000,000',
    country: { name: 'Japan', color: 'yellow' }
  }
];

describe('regroup tag', () => {
  it('parses arguments without output', () => {
    const result = djanjucks.renderString(
      '{% regroup cities by country as country_list %}',
      {
        cities: CITIES
      }
    );
    expect(result).toEqual('');
  });

  it('fails when missing "by" keyword', () => {
    expect(() => {
      djanjucks.renderString('{% regroup cities as country_list %}', {
        cities: CITIES
      });
    }).toThrow();
  });

  it('fails when missing "as" keyword', () => {
    expect(() => {
      djanjucks.renderString('{% regroup cities by country %}', {
        cities: CITIES
      });
    }).toThrow();
  });

  it('adds target to main context', () => {
    const result = djanjucks.renderString(
      `{% regroup cities by country as country_list %}
<ul>
{% for country in country_list %}
    <li>{{ country.grouper }}
    <ul>
        {% for city in country.list %}
          <li>{{ city.name }}: {{ city.population }}</li>
        {% endfor %}
    </ul>
    </li>
{% endfor %}
</ul>`,
      {
        cities: CITIES
      }
    );
    expect(result).toMatchSnapshot();
  });

  it('allows grouping with dot notation', () => {
    const result = djanjucks.renderString(
      `{% regroup cities by country.color as country_list %}

<ul>
{% for country in country_list %}
    <li>{{ country.grouper }}
    <ul>
        {% for city in country.list %}
          <li>{{ city.country.name }}: {{ city.name }}</li>
        {% endfor %}
    </ul>
    </li>
{% endfor %}
</ul>`,
      {
        cities: CITIES_WITH_OBJECT
      }
    );
    expect(result).toMatchSnapshot();
  });
});
