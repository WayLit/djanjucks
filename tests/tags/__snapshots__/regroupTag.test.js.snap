// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regroup tag adds target to main context 1`] = `
"
<ul>

    <li>India
    <ul>
        
          <li>Mumbai: 19,000,000</li>
        
          <li>Calcutta: 15,000,000</li>
        
    </ul>
    </li>

    <li>USA
    <ul>
        
          <li>New York: 20,000,000</li>
        
          <li>Chicago: 7,000,000</li>
        
    </ul>
    </li>

    <li>Japan
    <ul>
        
          <li>Tokyo: 33,000,000</li>
        
    </ul>
    </li>

</ul>"
`;

exports[`regroup tag allows grouping with dot notation 1`] = `
"

<ul>

    <li>red
    <ul>
        
          <li>India: Mumbai</li>
        
          <li>USA: Chicago</li>
        
    </ul>
    </li>

    <li>blue
    <ul>
        
          <li>India: Calcutta</li>
        
    </ul>
    </li>

    <li>yellow
    <ul>
        
          <li>USA: New York</li>
        
          <li>Japan: Tokyo</li>
        
    </ul>
    </li>

</ul>"
`;
